src/daft-algebra/src/simplify/boolean.rs

```diff
@@ -0,0 +1,284 @@
+    Ok(match expr.as_ref() {
+        Expr::BinaryOp { op, left, right } => match op {
+            // true AND e -> e
+            Operator::And if is_true(left) => Transformed::yes(right.clone()),
+            Operator::And if is_true(right) => Transformed::yes(left.clone()),
+            // false AND e -> false
+            Operator::And if is_false(left) => Transformed::yes(lit(false)),
+            Operator::And if is_false(right) => Transformed::yes(lit(false)),
+            // false OR e -> e
+            Operator::Or if is_false(left) => Transformed::yes(right.clone()),
+            Operator::Or if is_false(right) => Transformed::yes(left.clone()),
+            // true OR e -> true
+            Operator::Or if is_true(left) => Transformed::yes(lit(true)),
+            Operator::Or if is_true(right) => Transformed::yes(lit(true)),
+
+            Operator::And => simplify_and(expr.clone(), left, right),
+            Operator::Or => simplify_or(expr.clone(), left, right),
+
+            _ => Transformed::no(expr),
+        },
+        Expr::Not(not) => {
+            match not.as_ref() {
+                // !!e -> e
+                Expr::Not(not_not) => Transformed::yes(not_not.clone()),
+                // !true -> false, !false -> true
+                Expr::Literal(LiteralValue::Boolean(val)) => Transformed::yes(lit(!val)),
+                _ => Transformed::no(expr),
+            }
+        }
+        _ => Transformed::no(expr),
+    })
+}
+
+/// Combines common sub-expressions in AND expressions.
+///
+/// 1. Matches up each expression in the left-side conjunction with each expression in the right-side conjunction.
+///
+/// 2. For each pair, check if there are any common expressions in their disjunctions.
+///
+///     Ex:
+///     - left: (a OR b OR c OR ...)
+///     - right: (a OR b OR d OR ...)
+///     - common: a, b
+///
+/// 3. If there are common expressions, extract them out into ORs.
+///
+///     Ex (using above example):
+///     - a OR b OR ((c OR ...) AND (d OR ...))
+///
+/// 4. Finally, combine all new and remaining expressions back into conjunction
+fn simplify_and(expr: ExprRef, left: &ExprRef, right: &ExprRef) -> Transformed<ExprRef> {
+    let left_and_exprs = split_conjunction(left);
+    let right_and_exprs = split_conjunction(right);
+
+    let left_and_of_or_exprs = left_and_exprs
+        .iter()
+        .map(|e| split_disjunction(e).into_iter().collect())
+        .collect::<Vec<IndexSet<_>>>();
+    let right_and_of_or_exprs = right_and_exprs
+        .iter()
+        .map(|e| split_disjunction(e).into_iter().collect())
+        .collect::<Vec<IndexSet<_>>>();
+
+    let mut left_eliminated = vec![false; left_and_exprs.len()];
+    let mut right_eliminated = vec![false; right_and_exprs.len()];
+    let mut simplified_exprs = Vec::new();
+
+    for (i, left_exprs) in left_and_of_or_exprs.iter().enumerate() {
+        for (j, right_exprs) in right_and_of_or_exprs.iter().enumerate() {
```

Comment:
High level question: is there ever a case where we want to self-eliminate left exprs with left exprs (that are not itself)?

Reply:
Chatted offline. Since the optimizer rule is applied bottom up, we've already self-eliminated. LGTM